synapse/storage/databases/main/sliding_sync.py

```diff
@@ -0,0 +1,482 @@
+#
+# This file is licensed under the Affero General Public License (AGPL) version 3.
+#
+# Copyright (C) 2023 New Vector, Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# See the GNU Affero General Public License for more details:
+# <https://www.gnu.org/licenses/agpl-3.0.html>.
+#
+
+
+import logging
+from typing import TYPE_CHECKING, Dict, List, Mapping, Optional, Set, cast
+
+import attr
+
+from synapse.api.errors import SlidingSyncUnknownPosition
+from synapse.logging.opentracing import log_kv
+from synapse.storage._base import SQLBaseStore, db_to_json
+from synapse.storage.database import LoggingTransaction
+from synapse.types import MultiWriterStreamToken, RoomStreamToken
+from synapse.types.handlers.sliding_sync import (
+    HaveSentRoom,
+    HaveSentRoomFlag,
+    MutablePerConnectionState,
+    PerConnectionState,
+    RoomStatusMap,
+    RoomSyncConfig,
+)
+from synapse.util import json_encoder
+from synapse.util.caches.descriptors import cached
+
+if TYPE_CHECKING:
+    from synapse.storage.databases.main import DataStore
+
+logger = logging.getLogger(__name__)
+
+
+class SlidingSyncStore(SQLBaseStore):
+    async def persist_per_connection_state(
+        self,
+        user_id: str,
+        device_id: str,
+        conn_id: str,
+        previous_connection_position: Optional[int],
+        per_connection_state: "MutablePerConnectionState",
+    ) -> int:
+        """Persist updates to the per-connection state for a sliding sync
+        connection.
+
+        Returns:
+            The connection position of the newly persisted state.
+        """
+
+        store = cast("DataStore", self)
```

Comment:
Comment why cast is safe. This seems a little cheeky though if my comment is correct.

```suggestion
        # This cast is safe because the downstream code only cares about `store.get_id_for_instance(...)`
        # and `StreamWorkerStore` is mixed alongside `SlidingSyncStore` wherever we create a store.
        store = cast("DataStore", self)
```

Reply:
Yeah, its not great we're here. Mixins are awful, but here we are.

Reply:
(yes your comment is correct)

Reply:
Would it be bad to mixin `StreamWorkerStore` into this store to have more of a guarantee?

It's unclear if we should try to shy away from that sort of thing or not

Reply:
The really annoying this is that if this class depends on the other ones you rapidly hit problems with getting a consistent MRO ordering (and there are no automated tooling to help!). I've wasted hours before trying to fix MRO ordering, so not particularly keen on going down that road again.